common/preconditioner/isai_kernels.hpp.inc

```diff
@@ -46,94 +46,112 @@ template <int subwarp_size, int subwarps_per_block, typename ValueType,
+        }
 
-    if (row >= num_rows) {
-        return;
-    }
+        subwarp.sync();
 
-    const auto i_row_begin = i_row_ptrs[row];
-    const auto i_row_end = i_row_ptrs[row + 1] - i_row_begin < subwarp_size
-                               ? i_row_ptrs[row + 1]
-                               : i_row_begin + subwarp_size;
-    const auto num_row_elems = i_row_end - i_row_begin;
-
-    if (local_id < num_row_elems) {
-        const auto col = i_col_idxs[i_row_begin + local_id];
-        const auto m_row_end = m_row_ptrs[col + 1];
-        auto m_row_ptr = m_row_ptrs[col];
-        auto i_row_ptr = i_row_begin;
-        int idx{trisystem_start + local_id};  // write to trisystem in row major
-
-        while (m_row_ptr < m_row_end && i_row_ptr < i_row_end &&
-               idx < end_trisystem) {
-            const auto sparsity_col = i_col_idxs[i_row_ptr];
-            const auto m_col = m_col_idxs[m_row_ptr];
-            if (sparsity_col == m_col) {
-                trisystem[idx] = m_values[m_row_ptr];
-                idx += subwarp_size;
-                ++m_row_ptr;
-                ++i_row_ptr;
-            } else if (m_col < sparsity_col) {
-                ++m_row_ptr;
-            } else {
-                ++i_row_ptr;
-                idx += subwarp_size;
-            }
+        // Now, read a full col of `trisystem` into local registers, which will
+        // be row elements after this (implicit) transpose
+        ValueType local_row[subwarp_size];
+#pragma unroll
+        for (int i = 0; i < subwarp_size; ++i) {
+            local_row[i] = trisystem(i, local_id);
         }
-    }
-
-    this_block.sync();
-
-    // Now, read a full col of `trisystem` into local registers, which will be
-    // row elements after this (implicit) transpose
-    ValueType local_row[subwarp_size];
-    const auto trisystem_col_start = trisystem_start + local_id * subwarp_size;
-    for (int i = 0; i < subwarp_size; ++i) {
-        const auto intermediate_idx = local_id + i;
-        const auto coalesced_idx = intermediate_idx % subwarp_size;
-        local_row[coalesced_idx] =
-            trisystem[trisystem_col_start + coalesced_idx];
-    }
 
-    const auto rhs = trs_solve(num_row_elems, local_row, subwarp);
+        const auto rhs = trs_solve(i_row_size, local_row, subwarp);
 
-    // Write back:
-    if (local_id < num_row_elems) {
-        const auto idx = i_row_begin + local_id;
-        if (gko::isfinite(rhs)) {
-            i_values[idx] = rhs;
-        } else {
-            i_values[idx] =
-                i_col_idxs[idx] == row ? one<ValueType>() : zero<ValueType>();
+        // Write back:
+        if (local_id < i_row_size) {
+            const auto idx = i_row_begin + local_id;
+            if (is_finite(rhs)) {
+                i_values[idx] = rhs;
+            } else {
+                i_values[idx] = i_col_idxs[idx] == row ? one<ValueType>()
+                                                       : zero<ValueType>();
```

Comment:
```suggestion
                i_values[idx] = i_col_idxs[idx] == row;
```
does that works? or fail in complex

Reply:
In this case, I would like to be a bit more explicit, since otherwise the line doesn't immediately tell you what is going on.

Reply:
I see.